app.py

```diff
@@ -60,17 +95,69 @@ def show_in_filesystem(self, entity_type, entity_ids):
         with the given entity ids.
         """
         paths = []
-        
+
+        self.sgtk.synchronize_filesystem_structure()
+
         for eid in entity_ids:
+
+            if entity_type == "PublishedFile":
+                # Find the publish so that we can get the path back.
+                published_file = self.sgtk.shotgun.find_one("PublishedFile", [["id", "is", eid]], ['code', 'path'])
+                # Resolve the path for the local OS
+                try:
+                    local_path = sgtk.util.resolve_publish_path(self.sgtk, published_file)
+                except Exception as e:
+                    self.log_debug("Publish path couldn't be resolved falling back to context based path; reason: %s" % e)
+                    local_path = None
+
+                if local_path:
+                    if os.path.isdir(local_path):
+                        paths.append(local_path)
+                        continue
+                    elif os.path.isfile(local_path):
+                        paths.append(local_path)
+                        continue
+
+                    # possibly we have an image sequence and therefore its a symbolic path, instead check to see if
+                    # the parent folder of the path is valid and try opening that.
+                    # The issue with this logic is that possibly it was a directory that just didn't exist, as so
+                    # we would just be gathering the next directory up, ideally need a better way to handle sequences
+                    parent_dir = os.path.dirname(local_path)
+                    if parent_dir and os.path.isdir(parent_dir):
+                        paths.append(os.path.dirname(local_path))
+                        continue
+
             # Use the path cache to look up all paths linked to the task's entity
-            context = self.tank.context_from_entity(entity_type, eid)
-            paths.extend( context.filesystem_locations )
+            context = self.sgtk.context_from_entity(entity_type, eid)
+            context_paths = context.filesystem_locations
+
+            if context.step:
+                # As steps are non project entities and not linked from the step to the other entities you won't get step
+                # paths back from the context.filesystem_locations
+                # If we have a step in the context, we should check to see if we can resolve the path more deeply using that.
+                step_paths = self.sgtk.paths_from_entity("Step", context.step['id'])
+
+                for context_path in context_paths:
+                    # Loop over the context paths, and check to see if any of them can be extended with a step path.
+                    # If it can't fall back to the standard context path
+                    step_path = next((step_path for step_path in step_paths if step_path.startswith(context_path)),
```

Comment:
I don't use `next` much so I'm not exactly sure that this is doing. Is this taking the first "step path" that starts with `context_path` or something else?

Reply:
That's exactly right, and I provide a default value of `None` so that if it doesn't find a match then it won't error. It would be similar to writing

```
step_path = None
for a_step_path in step_paths:
    if a_step_path.startswith(context_path):
        step_path = a_step_path
        break
```

Reply:
Which is related to my question number 1, which is to say I'm assuming we should only care about one match. I'm not sure on the situations where context_paths might be multiple, and if we would ever potentially get two step path matches.